csrc/preseg_passes/reorder_sharded_axis.cpp

```diff
@@ -10,142 +10,178 @@
 #include <device_lower/utils.h>
 #include <fusion.h>
 #include <host_ir/lower.h>
+#include <ir/allocation_utils.h>
 #include <ir/base_nodes.h>
 #include <ir/interface_nodes.h>
 #include <ir/utils.h>
 #include <multidevice/utils.h>
 #include <ops/alias.h>
 #include <ops/arith.h>
+#include <scheduler/utils.h>
+#include <transform_replay.h>
 
 namespace nvfuser::preseg_passes {
 
+namespace {
+
+void transform_and_parallelize_like(TensorView* ref, TensorView* tv) {
+  auto old2new = reorderDIDToFront(ref);
+  propagateDIDTransform(
+      ref, {tv}, /*did_pos=*/old2new.size(), PropagateDirection::kForward);
+  shardAllLike(ref, {tv}, {kParallelTypeDIDs.begin(), kParallelTypeDIDs.end()});
+}
+
+void makeCommunicationLayoutCompliant(
+    Expr* expr,
+    CommunicationInfo communication_info) {
+  TensorView* input = expr->inputs().at(0)->as<TensorView>();
+  TensorView* output = expr->outputs().at(0)->as<TensorView>();
+
+  IterDomain* p_sharded_id = communication_info.p_sharded_id;
+  IterDomain* c_sharded_id = communication_info.c_sharded_id;
+
+  // Copy input if:
+  // 1. Input is not contiguous.
+  // 2. Input is not allocation compliant and is not a reduce/allreduce.
+  if (isTvContiguous(input) &&
+      (communication_info.type == CommunicationType::Reduce ||
+       isAllocationCompliant(input, p_sharded_id))) {
+    // If the input is already in the required memory layout,
+    // set its allocation explicitly to avoid changes by other passes.
+    input->setAllocationDomain(input->getMaybeAllocationDomain(), true);
+  } else {
+    // Copy input into required memory layout.
+    // Note: If input is not a fusion input, we should ideally be able to
+    // specify allocation domain of input instead. However, some schedulers
+    // (e.g. reduction) may not support this, so creating a copy here.
+    // This cannot always be done, for example, when this input is
+    // noncontiguous.
+
+    TensorView* input_copy = set(input);
+
+    // Find index of p_sharded_id in input_copy's logical domain.
+    // Reduction axis in input can change the position of p_sharded_id in
+    // input_copy's logical domain.
+    // This reordering is not needed for reduce/allreduce.
+    std::unordered_map<int64_t, int64_t> reorder_map = {};
+    if (communication_info.type != CommunicationType::Reduce) {
+      auto p2c =
+          PairwiseLogicalDomainMap(input, input_copy).mapProducerToConsumer();
+
+      IterDomain* input_copy_sharded_id = p2c.at(p_sharded_id);
+      int64_t input_copy_sharded_idx = posInDomain(
+          input_copy->getMaybeAllocationDomain(), input_copy_sharded_id);
+      reorder_map[input_copy_sharded_idx] = 0;
+    }
+
+    input_copy->setAllocationDomain(
+        TensorDomain::orderedAs(
+            input_copy->getMaybeAllocationDomain(), reorder_map),
+        true);
+
+    // New communication expression.
```

Comment:
The logic looks correct. Conceptually, you attempt to do the following

```
if cannot reuse input:
  input = input.cacheAfter
set input's allocation per communication

if cannot reuse output:
  output = output.cacheBefore
set output's allocation per communication

IrBuilder::create the new communication expression between input and output
```

I'm sure there are several details (e.g. reduction dimensions) that make your implementation much more complicated than that. I just need some more time to figure out a less convoluted way of doing that.

Reply:
I didn't look into using `cacheBefore/After`.
Let me try that.

Reply:
Using `cacheAfter/Before` will replace the `set` ops but much of the code around setting allocation domains remains.
One way to simplify this could be: https://github.com/NVIDIA/Fuser/pull/4256#discussion_r2103685625

Reply:
Recording offline discussion: the action order is
1. Fix the tests
2. Merge this PR
3. Merge #4554